fuel-vm/src/interpreter/memory.rs

```diff
@@ -359,6 +343,120 @@ impl MemoryInstance {
+
+        let heap_start = hp
+            .checked_sub(self.heap_offset())
+            .expect("Memory is invalid, hp is out of bounds");
+        let heap = &self.heap[heap_start..];
+        let desired_heap_start = hp
+            .checked_sub(desired_memory_state.heap_offset())
+            .expect("Memory is invalid, hp is out of bounds");
+        let desired_heap = &desired_memory_state.heap[desired_heap_start..];
+
+        let heap_changes = get_changes(heap, desired_heap, hp);
+
+        Some(MemoryRollbackData {
+            sp,
+            hp,
+            stack_changes,
+            heap_changes,
+        })
+    }
+
+    /// Rollbacks the memory changes returning the memory to the old state.
+    pub fn rollback(&mut self, data: &MemoryRollbackData) {
+        self.stack.resize(data.sp, 0);
+        assert!(
+            data.hp >= self.hp,
+            "We only allow shrinking of the heap during rollback"
+        );
+        self.hp = data.hp;
+
+        for change in &data.stack_changes {
+            self.stack[change.global_start
+                ..change.global_start.saturating_add(change.data.len())]
+                .copy_from_slice(&change.data);
+        }
+
+        let offset = self.heap_offset();
+        for change in &data.heap_changes {
+            let local_start = change
+                .global_start
+                .checked_sub(offset)
+                .expect("Invalid offset");
+            self.heap[local_start..local_start.saturating_add(change.data.len())]
+                .copy_from_slice(&change.data);
+        }
+    }
+}
+
+fn get_changes(
+    latest_array: &[u8],
+    desired_array: &[u8],
+    offset: usize,
+) -> Vec<MemorySliceChange> {
+    let mut changes = Vec::new();
+    let mut range = None;
+    for (i, (old, new)) in latest_array.iter().zip(desired_array.iter()).enumerate() {
+        if old != new {
+            range = match range {
+                None => Some((i, 1usize)),
+                Some((start, count)) => Some((start, count.saturating_add(1))),
+            };
+        } else if let Some((start, count)) = range.take() {
+            changes.push(MemorySliceChange {
+                global_start: offset.saturating_add(start),
+                data: desired_array[start..start.saturating_add(count)].to_vec(),
+            });
+        }
+    }
+    if let Some((start, count)) = range.take() {
+        changes.push(MemorySliceChange {
+            global_start: offset.saturating_add(start),
+            data: desired_array[start..start.saturating_add(count)].to_vec(),
+        });
+    }
+    changes
+}
+
+#[derive(Debug, Clone)]
+struct MemorySliceChange {
+    global_start: usize,
+    data: Vec<u8>,
+}
+
+/// The container for the data used to rollback memory changes.
+#[derive(Debug, Clone)]
+pub struct MemoryRollbackData {
+    /// Desired stack pointer.
+    sp: usize,
+    /// Desired heap pointer. Desired heap pointer can't be less than the current one.
```

Comment:
can't be equal or more ?

Reply:
Remember that heap only grows -> `HP` only decreases during `grow_heap`, so if we rollback to old state, `DESIRED_HP(or HP_BEFORE_GROW_HEAP) >= HP_AFTER_GROW_HEAP`